File: rules/pii/object-dump.go
package handler

import (
	"fmt"
	"log"
	"log/slog"

	"go.uber.org/zap"
)

type Request struct {
	UserID string
	Path   string
}

func logBad(r Request) {
	// ruleid: tqa-pii-object-dump-go
	log.Printf("handling request %+v", r)
}

func logBad2(r Request) {
	// ruleid: tqa-pii-object-dump-go
	msg := fmt.Sprintf("request: %+v", r)
	_ = msg
}

func logBad3(r Request, logger *zap.Logger) {
	// ruleid: tqa-pii-object-dump-go
	logger.Info("request", zap.Any("request", r))
}

func logGood(r Request) {
	// ok: tqa-pii-object-dump-go
	slog.Info("handling request", slog.String("path", r.Path))
}

func logGood2(status int) {
	// ok: tqa-pii-object-dump-go
	log.Printf("responded with status %d", status)
}

File: rules/pii/object-dump.py
import logging

logger = logging.getLogger(__name__)


class Request:
    def __init__(self, user_id, path):
        self.user_id = user_id
        self.path = path


def log_bad(req):
    # ruleid: tqa-pii-object-dump-py
    logger.info("handling request", req.__dict__)


def log_bad2(req):
    # ruleid: tqa-pii-object-dump-py
    logger.info("request", vars(req))


def log_bad3(request):
    # ruleid: tqa-pii-object-dump-py
    logger.info("incoming headers", request.headers)


def log_bad4(request):
    # ruleid: tqa-pii-object-dump-py
    logger.info("incoming body", request.body)


def log_good(req):
    # ok: tqa-pii-object-dump-py
    logger.info("handling request", extra={"path": req.path})


def log_good2(status):
    # ok: tqa-pii-object-dump-py
    logger.info("responded with status %s", status)